File: Tree/Balanced_Binary_Tree.cpp
#include <iostream>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        // if (root == NULL)
        if (!root)
            return true;
        int lh = cal_height(root->left);
        int rh = cal_height(root->right);
        if (abs(lh - rh) > 1)
            return false;
        else
            return isBalanced(root->left) && isBalanced(root->right);
    }

    int cal_height(TreeNode* node) {
        if (!node)
            return 0;
        int lh = cal_height(node->left);
        int rh = cal_height(node->right);
        return 1 + std::max(lh, rh);
    }
};

int main()
{
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(2);
    root->right = new TreeNode(4);
    root->left->left = new TreeNode(1);
    Solution s;
    int h = s.height(root);
    std::cout << h << "\n";
    return 0;
}

File: Tree/Balanced_Binary_Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isBalanced(self, root):
        if root is None:
            return True
        lh = self.cal_height(root.left)
        rh = self.cal_height(root.right)
        if abs(lh - rh) > 1:
            return False
        else:
            return self.isBalanced(root.left) and self.isBalanced(root.right)

    def cal_height(self, node):
        if node is None:
            return 0
        lh = self.cal_height(node.left)
        rh = self.cal_height(node.right)
        return 1 + max(lh, rh)
